### PR TITLE
Get exact char instead of last for /pay modifier; Added pay-modifier-enabled config line

### LINKED ISSUE #5495: Toggle /pay modifiers in config
### Feature description

I am suggesting adding a config line to enable/disable the `/pay` modifiers for 1 thousand, 1 million, 1 billion, and 1 trillion.
I already have this tested and ready to pr, but I read that this feature had to be requested first.

### How the feature is useful

This would, in my case, prevent players from scamming each other by convincing them that you can pay another currency to a player by adding the currency name afterwards.
For example. Our second currency is gems, so when a player runs `/pay 1gem` then they would pay the player 1,000,000 because the pay command checks for the last character and compares it with the available modifiers.
My PR would also fix this by checking for the first char after the number instead of the last.

### PR DESCRIPTION
**Pull request by AjMaacc**
### Information

This PR closes #5495 

### Details

**Proposed feature:**  

This feature allows the ability to enable/disable usage of /pay modifiers such as `1k`, `1m`, `1b` ,`1t`
Along with that when the command is checking for those modifiers, it checks the first char after the number instead of the last.

**Environments tested:**    

OS: Linux 

Java version:  17

- [x] Most recent Paper version (1.20.1, git-Paper-145)
- [ ] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8

**Demonstration:**

### CODE REVIEW

**Review by pop4959 — CHANGES_REQUESTED** [review]
Fundamentally this doesn't change anything, the same manipulation can be done with the first character as the last character. A proper fix is checking for _exactly_ the modifier, and if anything else is entered reject the input.

**Comment by pop4959** [comment]
Also worth considering if pay modifiers should even be enabled by default.

**Comment by AjMaacc** [comment]
> Fundamentally this doesn't change anything, the same manipulation can be done with the first character as the last character. A proper fix is checking for _exactly_ the modifier, and if anything else is entered reject the input.

I agree with only looking for the exact char. Would it be better to just ignore the modifier if it is not exact?

> Also worth considering if pay modifiers should even be enabled by default.

I'll have it default false

**Comment by pop4959** [comment]
> I agree with only looking for the exact char. Would it be better to just ignore the modifier if it is not exact?

I think it's safer to throw an exception there, and not assume the modifier (or no modifier) that the user intended to send if it's invalid. This shouldn't happen often, and showing an error will allow the user to make corrections and send the correct amount.

**Comment by AjMaacc** [comment]
Bump 💯

File: Essentials/src/main/java/com/earth2me/essentials/Settings.java [hunk 1]
```diff
@@ -1948,4 +1948,10 @@ public boolean showZeroBaltop() {
     public int getMaxItemLore() {
         return config.getInt("max-itemlore-lines", 10);
     }
+
+    @Override
+    public boolean isPayModifierEnabled() {
+        return config.getBoolean("pay-modifier-enabled", true);
```
Comment (pop4959) [inline Essentials/src/main/java/com/earth2me/essentials/Settings.java]: Should be false here as well if we want it to default to false if a new config isn't being generated

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandpay.java [hunk 2]
```diff
@@ -46,28 +46,32 @@ public void run(final Server server, final User user, final String commandLabel,
...
+                        tempAmount = tempAmount.multiply(TRILLION);
+                        break;
+                    }
+                    default: {
+                        throw new InvalidModifierException();
+                    }
+                }
             }
         }
-
```
Comment (pop4959) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandpay.java:70]: Try not to change formatting if possible such as adding/removing new lines in areas not related to your PR

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandpay.java [hunk 3]
```diff
@@ -135,4 +139,5 @@ protected List<String> getTabCompleteOptions(final Server server, final CommandS
             return Collections.emptyList();
         }
     }
+
```
Comment (pop4959) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandpay.java]: Same as above

File: Essentials/src/main/java/com/earth2me/essentials/commands/InvalidModifierException.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package com.earth2me.essentials.commands;
+
+import static com.earth2me.essentials.I18n.tl;
+
+/**
+ * Thrown when a player doesn't specify a currency modifier correctly.
+ */
+public class InvalidModifierException extends Exception{
```
Comment (pop4959) [inline Essentials/src/main/java/com/earth2me/essentials/commands/InvalidModifierException.java]:
```suggestion
public class InvalidModifierException extends Exception {
```

File: Essentials/src/main/resources/config.yml [hunk 5]
```diff
@@ -815,6 +815,11 @@ currency-symbol: '$'
 # For example, the euro symbol typically appears after the current amount.
 currency-symbol-suffix: false
 
+# Enable this to allow currency modifiers in /pay
+# (Ex. /pay Notch 1m = /pay Notch 1,000,000)
+# Available modfiers to use: k (Thousand), m (Million) b (Billion) t (Trillion)
```
Comment (pop4959) [inline Essentials/src/main/resources/config.yml]:
```suggestion
# Available modifiers: k (Thousand), m (Million) b (Billion) t (Trillion)
```

**Review by pop4959 — COMMENTED** [review]
Couple minor things, overall looks fine to me

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandpay.java (hunk 2 above)
Comment (AjMaacc) [reply Essentials/src/main/java/com/earth2me/essentials/commands/Commandpay.java:70]: This is essential for my pr to work

**Comment by AjMaacc** [comment]
Resolved Issues

File: Essentials/src/main/resources/messages.properties [hunk 6]
```diff
@@ -549,6 +549,7 @@ invalidHome=\u00a74Home\u00a7c {0} \u00a74doesn''t exist\!
 invalidHomeName=\u00a74Invalid home name\!
 invalidItemFlagMeta=\u00a74Invalid itemflag meta\: \u00a7c{0}\u00a74.
 invalidMob=\u00a74Invalid mob type.
+invalidModifier = \u00a74Invalid Modifier.
```
Comment (LaxWasHere) [inline Essentials/src/main/resources/messages.properties]:
```suggestion
invalidModifier=\u00a74Invalid Modifier.
```

**Comment by AjMaacc** [comment]
Bump 💯

**Pull request closed without merging**